### PR TITLE
Support user select use which GPU for mine

### PR DESCRIPTION
**Pull request by ShengguangXiao**
## Description

Let user select which GPU use for mine.

Add an GPU_TO_USE variable in constants.xml, let user choose which GPU used for mine.

https://github.com/Zilliqa/Zilliqa/issues/691

## Review Suggestion

Check file changed.

Find a machine with 2 GPUs, then change the constants.xml GPU_TO_USE option, and run ./build/tests/POW/Test_POW

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: constants.xml [hunk 1]
```diff
@@ -136,7 +136,8 @@
     -->
     </lookups>
     <gpu>
-        <NUM_DEVICE_TO_USE>1</NUM_DEVICE_TO_USE>
+        <!-- Which GPU to use, can use multile GPU, for example: 0, 2, 4 -->
```
Comment (ansnunez) [inline constants.xml]: Can you revise to `Which GPU to use, can use multiple GPU, for example: "0, 2, 4"`
(i.e., Fixed the misspelled "multiple" and added quotes in the example to make it clear it is the actual value you want to put in this field)

File: src/common/Constants.cpp [hunk 2]
```diff
@@ -78,6 +78,11 @@ unsigned int ReadGpuConstants(std::string propertyName) {
   return pt.get<unsigned int>("node.gpu." + propertyName);
 }
 
+std::string ReadGpuVariable(std::string propertyName) {
```
Comment (ansnunez) [inline src/common/Constants.cpp]: Can you change the parameter to `const std::string&` and the function name to `ReadGPUVariableFromConstantsFile`?

File: constants.xml (hunk 1 above)
Comment (ShengguangXiao) [reply constants.xml]: Ok

File: constants.xml [hunk 3]
```diff
@@ -136,7 +136,8 @@
     -->
     </lookups>
     <gpu>
-        <NUM_DEVICE_TO_USE>1</NUM_DEVICE_TO_USE>
+        <!-- Which GPU to use, can use multiple GPU, for example: "0, 2, 4" -->
```
Comment (ckyang) [inline constants.xml]: Maybe update the new GPU usage in wiki

File: src/libPOW/pow.cpp [hunk 4]
```diff
@@ -499,3 +510,14 @@ void POW::InitCUDA() {
...
 #endif
 }
+
+std::vector<unsigned int> POW::GetGpuToUse() {
+  std::vector<unsigned int> vecGpuIndex;
+  std::stringstream ss(GPU_TO_USE);
+  std::string item;
+  while (std::getline(ss, item, ',')) {
+    unsigned int index = strtol(item.c_str(), NULL, 10);
+    vecGpuIndex.push_back(index);
```
Comment (ckyang) [inline src/libPOW/pow.cpp]: Need to check duplicated value before inserting into container

File: constants.xml (hunk 3 above)
Comment (ShengguangXiao) [reply constants.xml]: OK

File: src/libPOW/pow.cpp [hunk 5]
```diff
@@ -474,21 +476,23 @@ void POW::InitCUDA() {
...
-
   if (!CUDAMiner::configureGPU(CUDA_BLOCK_SIZE, CUDA_GRID_SIZE, CUDA_STREAM_NUM,
                                CUDA_SCHEDULE_FLAG, 0, 0, false, false)) {
     LOG_GENERAL(FATAL, "Failed to configure CUDA GPU, please check hardware");
   }
 
   CUDAMiner::setNumInstances(UINT_MAX);
-  for (uint32_t i = 0; i < NUM_DEVICE_TO_USE; ++i) {
-    m_miners.push_back(std::make_unique<CUDAMiner>(i));
+  auto setGpuIndex = GetGpuToUse();
```
Comment (ckyang) [inline src/libPOW/pow.cpp]: The naming of variable should try to keep away of what the implementation it. Then, if someday we need to modify the implementation, we don't need to change the variable name again.

**Pull request merged**